src/Core/Grid/Query/CombinationQueryBuilder.php

```diff
@@ -0,0 +1,160 @@
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Grid\Query;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Query\QueryBuilder;
+use PrestaShop\PrestaShop\Core\Grid\Search\SearchCriteriaInterface;
+
+final class CombinationQueryBuilder extends AbstractDoctrineQueryBuilder
+{
+    /**
+     * @var DoctrineSearchCriteriaApplicatorInterface
+     */
+    private $searchCriteriaApplicator;
+
+    /**
+     * @param Connection $connection
+     * @param string $dbPrefix
+     * @param DoctrineSearchCriteriaApplicatorInterface $searchCriteriaApplicator
+     */
+    public function __construct(
+        Connection $connection,
+        string $dbPrefix,
+        DoctrineSearchCriteriaApplicatorInterface $searchCriteriaApplicator
+    ) {
+        parent::__construct($connection, $dbPrefix);
+        $this->searchCriteriaApplicator = $searchCriteriaApplicator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getSearchQueryBuilder(SearchCriteriaInterface $searchCriteria)
+    {
+        $qb = $this->getCombinationsQueryBuilder($searchCriteria)->select('pa.*');
+
+        $this->searchCriteriaApplicator
+            ->applyPagination($searchCriteria, $qb)
+            ->applySorting($searchCriteria, $qb)
+        ;
+
+        return $qb;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCountQueryBuilder(SearchCriteriaInterface $searchCriteria)
+    {
+        return $this->getCombinationsQueryBuilder($searchCriteria)
+            ->select('COUNT(pa.id_product_attribute)')
+        ;
+    }
+
+    /**
+     * @param SearchCriteriaInterface $searchCriteria
+     *
+     * @return QueryBuilder
+     */
+    private function getCombinationsQueryBuilder(SearchCriteriaInterface $searchCriteria): QueryBuilder
+    {
+        $filters = $searchCriteria->getFilters();
+        if (!isset($filters['product_id'])) {
+            //@todo: better exception? or require filter in another layer - like builder for filters? (categories has same issue)
+            throw new \RuntimeException('Product id is required for combinations grid');
+        }
+
+        $productId = (int) $filters['product_id'];
+
+        $qb = $this->connection->createQueryBuilder();
+        $qb->from($this->dbPrefix . 'product_attribute', 'pa')
+            ->where('pa.id_product = :productId')
+            ->setParameter('productId', $productId)
+        ;
+
+        // filter by attributes
+        if (isset($filters['attribute_ids'])) {
+            $combinationIds = $this->getCombinationIdsByAttributeIds($productId, (array) $filters['attribute_ids']);
```

Comment:
Is the cast array really useful?

Reply:
not sure if its right, but just in case we pass 1 value. (as we are used to pass one value to filter). You think I shiouldn't care about this?

Reply:
So if `$filters['attribute_ids'] = 42` for example does it cast it to `[42]` ?